gems/aws-sdk-core/lib/aws-sdk-core/plugins/checksum_algorithm.rb

```diff
@@ -49,122 +144,207 @@ def update(chunk)
+          when 'when_supported'
+            metrics << 'FLEXIBLE_CHECKSUMS_REQ_when_supported'
+          when 'when_required'
+            metrics << 'FLEXIBLE_CHECKSUMS_REQ_when_required'
+          end
+        end
+
+        def add_response_config_metric(config, metrics)
+          case config.response_checksum_validation
+          when 'when_supported'
+            metrics << 'FLEXIBLE_CHECKSUMS_RES_when_supported'
+          when 'when_required'
+            metrics << 'FLEXIBLE_CHECKSUMS_RES_when_required'
+          end
+        end
+
+        def add_request_checksum_metrics(algorithm, metrics)
+          case algorithm
+          when 'CRC32'
+            metrics << 'FLEXIBLE_CHECKSUMS_REQ_CRC32'
+          when 'CRC32C'
+            metrics << 'FLEXIBLE_CHECKSUMS_REQ_CRC32C'
+          when 'CRC64NVME'
+            metrics << 'FLEXIBLE_CHECKSUMS_REQ_CRC64'
+          when 'SHA1'
+            metrics << 'FLEXIBLE_CHECKSUMS_REQ_SHA1'
+          when 'SHA256'
+            metrics << 'FLEXIBLE_CHECKSUMS_REQ_SHA256'
+          end
+        end
+
+        def request_algorithm_selection(context)
+          return unless context.operation.http_checksum
+
+          input_member = context.operation.http_checksum['requestAlgorithmMember']
+          context.params[input_member.to_sym] ||= DEFAULT_CHECKSUM if input_member
+        end
+
+        def request_validation_mode(context)
+          return unless context.operation.http_checksum
+
+          input_member = context.operation.http_checksum['requestValidationModeMember']
+          context.params[input_member.to_sym] if input_member
+        end
+
+        def operation_response_algorithms(context)
+          return unless context.operation.http_checksum
+
+          context.operation.http_checksum['responseAlgorithms']
+        end
+
+        def checksum_required?(context)
+          (http_checksum = context.operation.http_checksum) &&
+            (checksum_required = http_checksum['requestChecksumRequired']) &&
+            (checksum_required && context.config.request_checksum_calculation == 'when_required')
+        end
+
+        def checksum_optional?(context)
           context.operation.http_checksum &&
-            (ChecksumAlgorithm.request_algorithm_selection(context) ||
-              context[:default_request_checksum_algorithm])
+            context.config.request_checksum_calculation == 'when_supported'
         end
 
-        def should_verify_response_checksum?(context)
-          context[:http_checksum][:validation_list] && !context[:http_checksum][:validation_list].empty?
+        def checksum_provided_as_header?(headers)
+          headers.any? { |k, _| k.start_with?('x-amz-checksum-') }
+        end
+
+        def should_calculate_request_checksum?(context)
+          !checksum_provided_as_header?(context.http_request.headers) &&
+            request_algorithm_selection(context) &&
+            (checksum_required?(context) || checksum_optional?(context))
```

Comment:
I believe there is an issue here with a user configures a client with an invalid `request_checksum_calculation`.  We don't do any validation of that when set on the client and then these two are mutually exclusive (and check only the known values).  

We could solve that in a few ways, but I'd vote for just a validation in client after init personally.  Alternatively, checksum_optional? could be changed to check `request_checksum_calculation != 'when_required'`